src/libcore/intrinsics.rs

```diff
@@ -962,59 +962,127 @@ extern "rust-intrinsic" {
+    /// Manually implement [`Vec::append`]:
     ///
     /// ```
-    /// use std::mem;
     /// use std::ptr;
     ///
-    /// # #[allow(dead_code)]
-    /// fn swap<T>(x: &mut T, y: &mut T) {
-    ///     unsafe {
-    ///         // Give ourselves some scratch space to work with
-    ///         let mut t: T = mem::uninitialized();
+    /// /// Moves all the elements of `src` into `dst`, leaving `src` empty.
+    /// fn append<T>(dst: &mut Vec<T>, src: &mut Vec<T>) {
+    ///     let src_len = src.len();
+    ///     let dst_len = dst.len();
     ///
-    ///         // Perform the swap, `&mut` pointers never alias
-    ///         ptr::copy_nonoverlapping(x, &mut t, 1);
-    ///         ptr::copy_nonoverlapping(y, x, 1);
-    ///         ptr::copy_nonoverlapping(&t, y, 1);
+    ///     // Ensure that `dst` has enough capacity to hold all of `src`.
+    ///     dst.reserve(src_len);
     ///
-    ///         // y and t now point to the same thing, but we need to completely forget `t`
-    ///         // because it's no longer relevant.
-    ///         mem::forget(t);
+    ///     unsafe {
+    ///         // The call to offset is always safe because `Vec` will never
+    ///         // allocate more than `isize::MAX` bytes.
+    ///         let dst_ptr = dst.as_mut_ptr().offset(dst_len as isize);
+    ///         let src_ptr = src.as_ptr();
+    ///
+    ///         // Truncate `src` without dropping its contents. We do this first,
+    ///         // to avoid problems in case something further down panics.
+    ///         src.set_len(0);
+    ///
+    ///         // The two regions cannot overlap becuase mutable references do
+    ///         // not alias, and two different vectors cannot own the same
+    ///         // memory.
+    ///         ptr::copy_nonoverlapping(src_ptr, dst_ptr, src_len);
+    ///
+    ///         // Notify `dst` that it now holds the contents of `src`.
+    ///         dst.set_len(dst_len + src_len);
     ///     }
     /// }
+    ///
+    /// let mut a = vec!['r'];
+    /// let mut b = vec!['u', 's', 't'];
+    ///
+    /// append(&mut a, &mut b);
+    ///
+    /// assert_eq!(a, &['r', 'u', 's', 't']);
+    /// assert!(b.is_empty());
     /// ```
+    ///
+    /// [`Vec::append`]: ../../std/vec/struct.Vec.html#method.append
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
 
-    /// Copies `count * size_of<T>` bytes from `src` to `dst`. The source
+    /// Copies `count * size_of::<T>()` bytes from `src` to `dst`. The source
     /// and destination may overlap.
     ///
-    /// `copy` is semantically equivalent to C's `memmove`.
+    /// If the source and destination will *never* overlap,
+    /// [`copy_nonoverlapping`] can be used instead.
+    ///
+    /// `copy` is semantically equivalent to C's [`memmove`], but with the argument
+    /// order swapped.  Copying takes place as if the bytes were copied from `src`
```

Comment:
To be semi-formal, this requires that `[src, src+len) ∪ [dst, dst+len)` is a memory range that is "separated from the heap" (which to me also includes a non-loopiness requirement).

Reply:
Uh. So how can we write this in a way that it is not confusing for the 99% of unsafe code that do not have to worry about different virtual addresses mapping to the same physical address?  I added this last sentence because @gnzlbg insisted. I am happy to remove it again...

>  this requires that [src, src+len) ∪ [dst, dst+len) is a memory range that is noalias

I have no idea what that *means*. Aliasing is usually a relational property -- this range is not allowed to alias with *what*?

Reply:
If you use 'loopy' memory, you're in entirely implementation-/un-defined behavior land. The rust model should not take this into account.

Reply:
> If you use 'loopy' memory, you're in entirely implementation-/un-defined behavior land. The rust model should not take this into account.

You don't want to be at the mercy of the compiler even if you use loopy memory.

Reply:
There is no sane way to resolve loopy memory at the model level, imo. You should absolutely be at the mercy of the compiler if you have two pointer values pointing to the same allocation in the same code.

Reply:
> There is no sane way to resolve loopy memory at the model level, imo. 

We would need to stop assuming that objects with different addresses are distinct which would certainly complicate many things. I am perfectly fine with leaving these uses of virtual memory as undefined behavior for now. 

> I think that removing the sentence is OK.

I would be more comfortable with explicitly calling the UB here out loud, something like: if the address ranges of `src` and `dst` are disjoint, but reads and writes behave as if these ranges would overlap (or something like that), the behavior is undefined.

Reply:
I would just point out the virtual memory stuff explicitly with an example; folks who read docs for these methods are likely to understand what is meant.

Reply:
I have many examples of this, but the following Linux-specific example is probably the easiest one to follow:

```rust
use libc::*;
// create an in-memory file
let mut fname = *b"/tmp/my_in_memory_fileXXXXXX\0";
let mut fd: c_long = syscall(SYS_memfd_create, fname.as_mut_ptr() as *mut c_char, 0);
assert_ne!(fd, -1);

// resize it to the memory page size
let fd = fd as c_int;  
let page_size = sysconf(_SC_PAGESIZE);
assert_ne!(ftruncate(fd, page_size as off_t), -1);

// Now we want to map the in-memory file to two adjacent memory regions. 
// This is done in two race-free steps. 
//
// First, allocate a large enough virtual memory region of `2 * page_size`,
// whose beginning refers to the contents of the in-memory file. The second
// half will refer to the in-memory file out-of-bounds after this operation.
let ptr = mmap(
    ptr::null_mut(),
    page_size * 2,
    PROT_READ | PROT_WRITE,
    MAP_SHARED,
    fd,
    0,
);
assert_ne!(ptr,  MAP_FAILED);

// Now we map the second half of the virtual memory region
// to the in-memory file as well, so that after this operation, 
// both the first half and the second half refer to the in-memory
// file
let ptr2 = mmap(
            (ptr as *mut u8).offset(page_size as isize) as *mut c_void,
            half_size,
            PROT_READ | PROT_WRITE,
            MAP_SHARED | MAP_FIXED,
            fd,
            0,
);
assert_ne!(ptr2, MAP_FAILED);
        
// now we have created loopy-memory
let ptr = ptr as *mut u8;
// such that this is undefined behavior because we are
// copying memory to disjoint virtual memory addresses
// that refer to the same physical memory and therefore
// overlap: 
ptr::copy_nonoverlapping(ptr, ptr.offset(page_size)); // UB
```

Reply:
That's not very easy to follow though, and 99% of the users won't even care -- worse, many of them will be very confused.

This is far from the only place where things break if different virtual addresses map to the same location, is it not? TBH I think this needs a proper review for what we can still guarantee, not some hastily done local change.

I get the feeling that just because I dared touching these docs at all, now everything that was ever wrong with them must be fixed at once. That's kind of an unfair standard, and it would take more time than I am willing to put into this.

Reply:
> That's not very easy to follow though, and 99% of the users won't even care

I agree and I don't think we should include it. @Centril asked for an example, and AFAIK that's the easiest example that shows how to do this and that people familiar with mmap might be able to understand. Every other OS API produces examples that are much much harder to understand than this one. 

> This is far from the only place where things break if different virtual addresses map to the same location, is it not? 

I'd guess so. I was actually expecting the code below using asserts without atomic or volatile loads to actually be misoptimized by llvm because I am reading from the same pointer twice without modifying the memory behind the pointer.

In any case, I mentioned above that I would be more comfortable if the text would call the UB explicitly:

> if the address ranges of src and dst are disjoint, but reads and writes behave as if these ranges would overlap (or something like that), the behavior is undefined.

but I think it would be better to just open an issue about this where this can be discussed independently of this PR. We should just merge this, and keep improving it incrementally once it is clearer how to do so.

Reply:
> if the address ranges of src and dst are disjoint, but reads and writes behave as if these ranges would overlap (or something like that), the behavior is undefined.

I am pretty sure that would leave people *really* confused.

The documentation as it is now in this PR is pretty much equivalent to [the one for `memmove`](https://en.cppreference.com/w/c/string/byte/memmove). If it is good enough for C to not give any thought to virtual memory effects, I think it is good enough for us *for now*. ;)

Reply:
Moved to https://github.com/rust-rfcs/unsafe-code-guidelines/issues/28